### PR TITLE
refactor!: black-box integration tests

### LINKED ISSUE #4136: [suggestion] Refactor Iroha CLI
## Description

Apart from not using `clap`, the CLI (`iroha` binary) has other issues:

- It is simultaneously a binary and a library, which looks confusing IMO. It is this way so `test_network` crate can re-use `Iroha` root orchestrator in test env
- As a library, it exports `iroha::samples` with data samples for tests. Doesn't seem to relate to CLI at all.
- `Torii` is defined here; again, it is not really related to CLI and rather deserves to be moved somewhere.

To address this issues and make another steps towards the Configuration Overhaul RFC (https://github.com/hyperledger/iroha/issues/2585), I propose the following actions.

### Actions

- [x] Remove `Torii` from `iroha`; make it a separate `iroha_torii` crate. It is a self-contained module, focusing on Iroha HTTP API. While _core_ enough to belong to `iroha_core` (like Sumeragi or Kura), IMO it would better be a separate crate because its peripheral role, and it's a good candidate to be a standalone compilation unit.
  - #4139
- [x] Remove global panic hook setting from `Iroha` startup; set it up on a higher level (e.g. CLI)
  - Addressed by #4716 
- [ ] Remove `iroha::samples` from `iroha`; maybe, make a separate `test_samples` crate or something. 
- [ ] Consider moving `Iroha` structure & logic out of CLI, e.g. into `iroha_orchestrator` or something.
- [x] https://github.com/hyperledger/iroha/issues/4105 (+ bump `clap` version of `iroha_client_cli`)
- [x] Additionally, remove setting genesis file through CLI and introduce `genesis.file` configuration parameter as outlined in https://github.com/hyperledger/iroha/issues/4029

### LINKED ISSUE #4383: Refactor config usage in integration tests
Namely, `impl TestConfig for Config` and `iroha::samples::get_user_config`

https://github.com/hyperledger/iroha/blob/e1e8a1a3bc3410ba14d34e8c9dd089a1a8817016/core/test_network/src/lib.rs#L766

https://github.com/hyperledger/iroha/blob/50b5bc2af3198a39a045e174219161116482ef15/cli/src/samples.rs#L62

The use of configuration in tests is limited and unlikely to require building the user layer of it. As configuration validation logic (e.g. validating that `genesis.file` exists), these parts start to fall.

To be less abstract, I propose to refactor these parts to use the config in a way that is valid in tests and will not rely on what is not designed for tests. For example, it doesn't make sense to specify `genesis.file` in a test config when it will be never read there.

Related:

- #4136

### LINKED ISSUE #4500: Proper black-box integration tests MVP
We've long had an issue that our `client/tests/integration` are not proper black box tests. This means the have `iroha` as dependency  when they shouldn't.

The main goal is to remove `iroha` and `test_network` as `dev-dependencies` of `iroha_client`. However, the scope of this task is to produce an MVP with only several tests migrated to this new scheme of running integration tests. Later we can migrate all of the tests, but it's important to provide a framework for new tests first.

We should use [this](https://docs.rs/testcontainers/latest/testcontainers/) to bring up dockers running iroha for every test.

### PR DESCRIPTION
**Pull request by 0x009922**
_Re-opening of #5087_

Build on top of, and should be undrafted after merging of:

- [x] #5121 
- [x] #5122 
- [x] #5123 

---

## Context

`iroha_test_network` has a few problems:

- It doesn't test Iroha as a black box, but runs directly using internal implementation details.
- Execution flow of peers is messy, hard to debug & control. (e.g. for cases of restarts). Logs from all peers are all over the place, very noisy.
- Messy API: hard to fine tune the network & peers for specific cases; high coupling with defaults; `thread::sleep`s with pipeline time instead of providing precise lifecycle hooks
- Non-optimal defaults: most of the tests don't need 4 peers with 4s of pipeline time. A single peer with close-to-zero block time works significantly faster.

Black boxing would mean to run Iroha through its CLI as a dedicated process, just as users would do.

### Solution

Re-implement `iroha_test_network`:

- Black-boxing: run `irohad` as a direct child process. Close #4500 
  - Create temp dir for every randomly generated peer and store there its state, configs, and logs.
  - Configuration through raw TOML - feel your users! Close #4383 
  - Can work with any `irohad` target (e.g. `debug`, `release`)
- Allocate ports automatically (`fslock`-based solution) - no more manual ports setting. Works intra-process (`cargo test`) and inter-process (`cargo nextest` #4987).
- Higher-level API suitable for writing expressive tests.
- Optimal defaults: a single peer with very short timings.
- Faster!
- Bonus: test execution can now be terminated immediately with `SIGINT` (Ctrl + C). It used to be suspended.

Other changes:

- Rewrite many tests, reduce flakiness to the minumum. Turns out some were broken, but passing for other reasons. It was primarily a cause of a messy test network API. Partially address #4516
- Remove `unstable_network` tests: they rely on a direct violation of black boxing - the use of `FreezeStatus` to make peers faulty. To be re-implemented in some other way. 
- Make `irohad` a closed binary; don't expose `Iroha`; remove samples from it. Close #4136
- Minor changes in `iroha_core`.

### Further steps

- #3130 is a must
- Make `iroha_test_network` an executable, use in pytests and in SDKs
- **Move tests of business logic** from integration to `iroha_core` itself.

### Flaky tests

- `restarted_peer_should_have_the_same_asset_amount` - possibly due to a bug in Iroha? Read FIXME comment there.
- `extra_functional::connected_peers::*` _sometimes_ fail due to #5104 

### Migration Guide

**For devs:** you need to have `irohad` available in `$PATH` before running tests using `iroha_test_network`. The easiest way to achieve this is by running:

```sh
cargo install --path crates/irohad --locked
```

---

### Review notes

Due to Client still being blocking, there is some ugly code with `spawn_blocking`.

### Checklist

- [x] I've read [`CONTRIBUTING.md`](../CONTRIBUTING.md).
- [x] (optional) I've written unit tests for the code changes.
- [x] All review comments have been resolved.
- [x] All CI checks pass.
- [x] Upload test network artifacts

### CODE REVIEW

**Comment by 0x009922** [comment]
`test_network_runs` artifacts takes 70mb. Is this okay?

File: crates/iroha_crypto/src/lib.rs [hunk 1]
```diff
@@ -774,6 +774,11 @@ impl PrivateKey {
     pub fn to_bytes(&self) -> (Algorithm, Vec<u8>) {
         (self.algorithm(), self.payload())
     }
+
+    /// Wrap itself into [`ExposedPrivateKey`].
+    pub fn expose(self) -> ExposedPrivateKey {
+        ExposedPrivateKey(self)
+    }
```
Comment (0x009922) [inline crates/iroha_crypto/src/lib.rs:781]: Not really a necessary change, can remove.

File: crates/iroha_torii/src/lib.rs
Comment (0x009922) [inline crates/iroha_torii/src/lib.rs:1]: Extra change, will revert.

File: crates/iroha_torii/src/utils.rs [hunk 2]
```diff
@@ -64,6 +64,7 @@ pub mod extractors {
     }
 
     /// Extractor of Accept header
+    #[allow(unused)] // unused without `telemetry` feature
```
Comment (0x009922) [inline crates/iroha_torii/src/utils.rs:67]: Shall use `cfg_attr` instead

File: crates/iroha_wasm_builder/src/lib.rs [hunk 3]
```diff
@@ -361,6 +361,7 @@ impl Output {
 }
 
 // TODO: Remove cargo invocation (#2152)
+#[allow(unreachable_code, unused)]
```
Comment (0x009922) [inline crates/iroha_wasm_builder/src/lib.rs:364]: Will remove.

File: hooks/pre-commit.sample [hunk 4]
```diff
@@ -3,23 +3,22 @@
...
+cargo fmt --manifest-path ./wasm_samples/Cargo.toml --all -- --check
+# lints
+cargo clippy --workspace --benches --tests --examples --all-features
+# TODO: fails, re-enable
+# cargo clippy --workspace --benches --tests --examples --no-default-features
+# build WASMs
+./scripts/build_wasm_samples.sh
 # update the default executor
-cargo run --release --bin iroha_wasm_builder -- build ./wasm_samples/default_executor --optimize --out-file ./defaults/executor.wasm
+cp ./wasm_samples/target/prebuilt/default_executor.wasm ./defaults/executor.wasm
```
Comment (0x009922) [inline hooks/pre-commit.sample:14]: Extra line; it is copied by the script above.

File: hooks/pre-commit.sample
Comment (0x009922) [inline hooks/pre-commit.sample:1]: Code be split into a separate PR.

File: pytests/iroha_cli_tests/test/assets/test_register_asset_definitions.py [hunk 5]
```diff
@@ -105,6 +105,7 @@ def test_register_asset_with_not_existing_domain(
         iroha_cli.should(have.error(Stderr.FAILED_TO_FIND_DOMAIN.value))
 
 
+# FIXME: this should not work anymore
```
Comment (0x009922) [inline pytests/iroha_cli_tests/test/assets/test_register_asset_definitions.py:108]: @AlexStroke this test is broken - Iroha no longer enforces restrictions on asset names.

File: crates/iroha/tests/integration/queries/mod.rs [hunk 6]
```diff
@@ -14,9 +14,11 @@ mod role;
 mod smart_contract;
 
 #[test]
+// FIXME
+#[ignore = "started to fail after #5086?"]
```
Comment (0x009922) [inline crates/iroha/tests/integration/queries/mod.rs:18]: Started to fail after #5086 - will need to look at this test again.

File: crates/iroha/tests/integration/set_parameter.rs [hunk 7]
```diff
@@ -12,8 +12,10 @@ use iroha_test_network::*;
 
 #[test]
 fn can_change_parameter_value() -> Result<()> {
-    let (_rt, _peer, test_client) = <PeerBuilder>::new().with_port(11_135).start_with_runtime();
-    wait_for_genesis_committed(&vec![test_client.clone()], 0);
+    let (network, _rt) = NetworkBuilder::new()
+        .with_default_pipeline_time()
```
Comment (0x009922) [inline crates/iroha/tests/integration/set_parameter.rs:16]: No need for this.

Will optimise this test to use some other parameter to check for.

File: crates/iroha/tests/integration/permissions.rs [hunk 8]
```diff
@@ -13,51 +13,39 @@ use iroha_executor_data_model::permission::{
...
 };
-use iroha_genesis::GenesisBlock;
-use iroha_test_network::{PeerBuilder, *};
+use iroha_test_network::*;
 use iroha_test_samples::{gen_account_in, ALICE_ID, BOB_ID};
+use tokio::{join, time::timeout};
 
-#[test]
-fn genesis_transactions_are_validated_by_executor() {
+// FIXME
```
Comment (0x009922) [inline crates/iroha/tests/integration/permissions.rs:20]: Forgot why this failed, will investigate.

File: crates/iroha/tests/integration/extra_functional/normal.rs
Comment (0x009922) [inline crates/iroha/tests/integration/extra_functional/normal.rs:1]: Don't know what is `extra_functional` about this one. It is really _normal_.

File: crates/iroha/tests/integration/extra_functional/mod.rs
Comment (0x009922) [inline crates/iroha/tests/integration/extra_functional/mod.rs:1]: Shall we get rid of `extra_functional` and move these tests to the top level?

@mversic

**Review by 0x009922 — COMMENTED** [review]
Self-review

File: crates/iroha_torii/src/lib.rs
Comment (0x009922) [reply crates/iroha_torii/src/lib.rs:1]: Not extra, has to do with warning fixes

File: crates/iroha/tests/integration/extra_functional/mod.rs
Comment (Erigara) [reply crates/iroha/tests/integration/extra_functional/mod.rs:1]: @AlexStroke do you remember why it was introduced?

**Review by s8sato — COMMENTED** [review]
Reviewed overall except for `crates/iroha/tests` for now. I think this is a great job of making it faster as well as more reliable as integration tests

File: .github/workflows/iroha2-dev-pr.yml [hunk 9]
```diff
@@ -92,94 +94,43 @@ jobs:
...
-      - name: Run integration tests, with all features
-        run: >
-          mold --run cargo nextest run 
-          --all-features
-          --no-fail-fast 
-          --failure-output final 
-          --test-threads 1
-          -E 'test(extra_functional)'
+          name: test_network_runs
+          path: ${{ env.TEST_NETWORK_TMP_DIR }}/irohad_test_network_*
```
[the remark this answers is not in the document]
Comment (0x009922) [reply .github/workflows/iroha2-dev-pr.yml:133]: Peer logs aren't printed during tests, except for STDERR (but not always). They are too verbose and it is much more convenient to inspect them stored in separate files, for separate peers.

So, these artifacts contain each generated peer's configurations, logs, and even Kura storage. If some test fails, it's possible to fetch the artifacts, find a specific directory, and inspect what was going on with a specific peer.

File: crates/iroha/Cargo.toml [hunk 10]
```diff
@@ -83,37 +83,15 @@ toml = { workspace = true }
...
-
-[[bench]]
-name = "tps-dev"
-harness = false
-path = "benches/tps/dev.rs"
-
-[[example]]
-name = "tps-oneshot"
-harness = false
-path = "benches/tps/oneshot.rs"
```
[the remark this answers is not in the document]
Comment (0x009922) [reply crates/iroha/Cargo.toml:119]: I agree on that. But that's too much work for this PR to rewrite them.

Created https://github.com/hyperledger/iroha/issues/5150

File: crates/iroha_test_network/src/lib.rs [hunk 11]
```diff
@@ -1,852 +1,943 @@
...
-    .clone()
+    PATH.get_or_init(|| {
+        if let Ok(path) = which::which("irohad") {
+            path
+        } else {
+            eprintln!(
+                "ERROR: could not locate `irohad` binary in $PATH\n  \
+                It is required to run `iroha_test_network`.\n  \
+                The easiest way to satisfy this is to run:\n\n    \
+                cargo install ./crates/irohad --locked"
```
[the remark this answers is not in the document]
Comment (0x009922) [reply crates/iroha_test_network/src/lib.rs:69]: Thanks for the catch.

File: hooks/pre-commit.sample [hunk 12]
```diff
@@ -3,23 +3,20 @@
...
-cd -
-# update the default executor
-cargo run --release --bin iroha_wasm_builder -- build ./wasm_samples/default_executor --optimize --out-file ./defaults/executor.wasm
+cargo fmt --manifest-path ./wasm_samples/Cargo.toml --all -- --check
+# lints
+cargo clippy --workspace --benches --tests --examples --all-features
+# TODO: fails, re-enable
+# cargo clippy --workspace --benches --tests --examples --no-default-features
+# build WASMs
+./scripts/build_wasm_samples.sh
```
[the remark this answers is not in the document]
Comment (0x009922) [reply hooks/pre-commit.sample:12]: You are right. I'd better remove it from here entirely.

File: crates/iroha/tests/integration/events/pipeline.rs [hunk 13]
```diff
@@ -1,134 +1,105 @@
...
-                .expect("Failed to create event iterator.");
-            let event_result = event_iterator.next().expect("Stream closed");
-            let _event = event_result.expect("Must be valid");
-        })
-    }
+    Ok(())
 }
 
 #[test]
+#[ignore = "unclear how to test it while treating Iroha as a block box"]
```
[the remark this answers is not in the document]
Comment (0x009922) [reply crates/iroha/tests/integration/events/pipeline.rs:78]: Agreed. Don't want to create an issue for it though. Would leaving a comment in code be fine?

**Comment by 0x009922** [comment]
#5140 caused regression in `extra_functional::genesis` tests (see https://github.com/hyperledger/iroha/pull/5140#issuecomment-2415594233). Not sure yet if I need to rewrite tests.

File: .github/workflows/iroha2-dev-pr.yml (hunk 9 above)
[the remark this answers is not in the document]
Comment (s8sato) [reply .github/workflows/iroha2-dev-pr.yml:133]: If artifact size is a concern, we can consider tweaking `retention-days` and `compression-level`

File: crates/iroha/tests/integration/extra_functional/unstable_network.rs
Comment (s8sato) [inline crates/iroha/tests/integration/extra_functional/unstable_network.rs:1]: Any tracking issue to re-implement tests with faulty peers? Btw its role can be partially overlapped with `extra_functional::offline_peers`

File: crates/iroha/tests/integration/events/data.rs [hunk 14]
```diff
@@ -124,105 +127,84 @@ fn wasm_execution_should_produce_events() -> Result<()> {
...
+    let network = NetworkBuilder::new().start().await?;
+    let mut events_stream = network
+        .client()
+        .listen_for_events_async([DataEventFilter::Any])
+        .await?;
+
+    {
+        let client = network.client();
+        let tx = client.build_transaction(executable, <_>::default());
+        spawn_blocking(move || client.submit_transaction_blocking(&tx)).await??;
```
Comment (s8sato) [inline crates/iroha/tests/integration/events/data.rs:146]: This seems to be a common pattern. Is the goal to delegate blocking tasks to the dedicated thread pool to reduce overall test time?

It is fine to me if this is a workaround until #3130 async client. Otherwise, developers may not want to be aware of concurrency with other tests and may forget to wrap a blocking task in `spawn_blocking`

Comment (s8sato) [reply crates/iroha/tests/integration/events/data.rs:146]: I just read the review notes now

Comment (0x009922) [reply crates/iroha/tests/integration/events/data.rs:146]: > I just read the review notes now

Yeah (

It is the best temporary solution I found until the async client is there. Though, I could have introduced some way to reduce boilerplate of `clone`+`spawn_blocking`+`move`, like an extension trait to the `Client`, I thought it's better to keep tests very simple and rely on `tokio` utils directly as much as possible.

File: crates/iroha/tests/integration/extra_functional/unstable_network.rs
Comment (0x009922) [reply crates/iroha/tests/integration/extra_functional/unstable_network.rs:1]: Here it is: https://github.com/hyperledger/iroha/issues/5153

File: crates/iroha/tests/integration/extra_functional/offline_peers.rs [hunk 15]
```diff
@@ -56,22 +74,24 @@ fn register_offline_peer() -> Result<()> {
...
 
     Ok(())
 }
 
-fn check_status(peer_clients: &[Client], expected_blocks: u64) {
-    let n_peers = peer_clients.len() as u64;
-
-    for peer_client in peer_clients {
-        let status = peer_client.get_status().unwrap();
+async fn check_status(network: &Network, expected_peers: u64) {
```
[the remark this answers is not in the document]
Comment (0x009922) [reply crates/iroha/tests/integration/extra_functional/offline_peers.rs:87]: I hesitate to extend `Network` with `ensure_peers`. `ensure_blocks` works differently, relying on the events stream set up for each peer. This raises questions whether to implement peers tracking in the same manner, use or not polling, and so on...

This small function seems to do the job, for now.

**Comment by 0x009922** [comment]
@BAStos525 please take a look at the CI changes.

File: .github/workflows/iroha2-dev-pr.yml (hunk 9 above)
[the remark this answers is not in the document]
Comment (0x009922) [reply .github/workflows/iroha2-dev-pr.yml:133]: Set retention days to 3. Not sure if compression level will help much, as it seems the most space is occupied by the binary blobs of `genesis.scale` with pre-optimised WASMs inside.

File: crates/iroha_test_network/src/lib.rs [hunk 16]
```diff
@@ -1,852 +1,943 @@
...
-        for &peer in &peers[0..n_genesis_peers] {
-            result[peer] = PeerInfo::Online { is_genesis: true };
-        }
-        for &peer in peers.iter().rev().take(n_offline_peers) {
-            result[peer] = PeerInfo::Offline;
-        }
-        result
+    pub fn consensus_estimation(&self) -> Duration {
+        self.block_time + self.commit_time / 2
     }
```
[the remark this answers is not in the document]
Comment (0x009922) [reply crates/iroha_test_network/src/lib.rs]: Replaced.

File: pytests/iroha_cli_tests/test/assets/test_register_asset_definitions.py (hunk 5 above)
Comment (0x009922) [reply pytests/iroha_cli_tests/test/assets/test_register_asset_definitions.py:108]: @alexstroke1

**Review by nxsaken — DISMISSED** [review]

File: .github/workflows/iroha2-dev-pr.yml (hunk 9 above)
[the remark this answers is not in the document]
Comment (nxsaken) [reply .github/workflows/iroha2-dev-pr.yml:133]: Would it make sense to only upload test artifacts if tests fail?

[the remark this answers is not in the document]
Comment (s8sato) [reply .github/workflows/iroha2-dev-pr.yml:133]: @0x009922 

We could assume developers should review cause of the failure in 1 day, but I'd not insist on this.

Rather, `run-*-genesis.scale` for each run of each test looks redundant to me. I think only few tests are modifying the default test genesis. So each run could record some reference to genesis

[the remark this answers is not in the document]
Comment (0x009922) [reply .github/workflows/iroha2-dev-pr.yml:133]: > Would it make sense to only upload test artifacts if tests fail?

Makes sense. 

> Rather, `run-*-genesis.scale` for each run of each test looks redundant to me. I think only few tests are modifying the default test genesis. So each run could record some reference to genesis

There is only one `genesis.scale` for each peer, it doesn't vary between runs. Plus, there are very few tests that restart peers. But anyways, while I totally agree that duplication of genesis is sub-optimal, implementing optimisation with a reference seems to be a premature optimisation to me.

~~Let me set the max compression level - maybe it will be smart-enough to de-duplicate `genesis.scale`?~~ I just remembered that each genesis contains trusted peers encoded, which is a unique value each time, so it's no use.

File: crates/iroha_torii/src/utils.rs [hunk 17]
```diff
@@ -64,6 +64,7 @@ pub mod extractors {
     }
 
     /// Extractor of Accept header
+    #[cfg_attr(not(feature = "telemetry"), allow(unused))]
```
Comment (mversic) [inline crates/iroha_torii/src/utils.rs:67]:
```suggestion
    #[cfg_attr(not(feature = "telemetry"), expect(unused))]
```

File: .github/workflows/iroha2-dev-pr.yml [hunk 18]
```diff
@@ -92,94 +94,44 @@ jobs:
...
-      - uses: actions/checkout@v4
-      - name: Download executor.wasm
-        uses: actions/download-artifact@v4
-        with:
-          name: executor.wasm
-          path: ${{ env.DOCKER_COMPOSE_PATH }}
-      - name: Download the rest of WASM samples
-        uses: actions/download-artifact@v4
+      - name: Upload test network artifacts
+        if: failure()
```
Comment (nxsaken) [inline .github/workflows/iroha2-dev-pr.yml:129]: This will run if any step fails for any reason, I think. I wonder if it's possible to run this only when `cargo nextest` returns a specific code for test failure

Comment (BAStos525) [reply .github/workflows/iroha2-dev-pr.yml:129]: Am I right that test network artifacts should be uploaded when `Test with no default features` and/or `Test with all features` steps are failed?

Comment (nxsaken) [reply .github/workflows/iroha2-dev-pr.yml:129]: Yes, do you see any problems with that?

Comment (s8sato) [reply .github/workflows/iroha2-dev-pr.yml:129]: I think we can [add conditions](https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/evaluate-expressions-in-workflows-and-actions#failure-with-conditions) to upload only when specific steps failed

Comment (BAStos525) [reply .github/workflows/iroha2-dev-pr.yml:129]: Might be
```
      - name: Test with no default features
        id: no_feat_tests
        run: >
          mold --run cargo llvm-cov nextest
          --no-default-features
          --branch --no-report

      - name: Test with all features
        id: all_feat_tests
        run: >
          mold --run cargo llvm-cov nextest
          --all-features
          --branch --no-report
      - name: Upload test network artifacts
        if: ${{ (failure() && steps.no_feat_tests.outcome == 'failure') || (failure() && steps.all_feat_tests.outcome == 'failure') }}
        uses: actions/upload-artifact@v4
        with:
          name: test_network_runs
          path: ${{ env.TEST_NETWORK_TMP_DIR }}/irohad_test_network_*
          retention-days: 3
```

Comment (0x009922) [reply .github/workflows/iroha2-dev-pr.yml]: Updated the policy accordingly

**Review by nxsaken — DISMISSED** [review]

File: pytests/iroha_cli_tests/test/assets/test_register_asset_definitions.py (hunk 5 above)
Comment (0x009922) [reply pytests/iroha_cli_tests/test/assets/test_register_asset_definitions.py:108]: Yes, they aren't failing, while they should - an assertion that a long account cannot be registered would fail.

Comment (mversic) [reply pytests/iroha_cli_tests/test/assets/test_register_asset_definitions.py:108]: make an issue and resolve discussion

Comment (0x009922) [reply pytests/iroha_cli_tests/test/assets/test_register_asset_definitions.py:108]: Created https://github.com/hyperledger/iroha/issues/5161

**Review by mversic — DISMISSED** [review]

**Review by mversic — APPROVED** [review]

**Review by nxsaken — APPROVED** [review]

**Review by s8sato — APPROVED** [review]
